font/src/directwrite/mod.rs

```diff
@@ -13,30 +13,168 @@
+    collection: FontCollection,
+    fallback: Option<FontFallback>,
+}
+
+impl DirectWriteRasterizer {
+    fn rasterize_glyph(
+        &self,
+        face: &FontFace,
+        size: Size,
+        c: char,
+    ) -> Result<RasterizedGlyph, Error> {
+        let glyph_index = self.get_glyph_index(face, c)?;
+
+        let em_size = em_size(size);
+
+        let glyph_run = DWRITE_GLYPH_RUN {
+            fontFace: unsafe { face.as_ptr() },
+            fontEmSize: em_size,
+            glyphCount: 1,
+            glyphIndices: &glyph_index,
+            glyphAdvances: &0.0,
+            glyphOffsets: &GlyphOffset::default(),
+            isSideways: 0,
+            bidiLevel: 0,
+        };
+
+        let rendering_mode = face.get_recommended_rendering_mode_default_params(
+            em_size,
+            self.device_pixel_ratio,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+        );
+
+        let glyph_analysis = GlyphRunAnalysis::create(
+            &glyph_run,
+            self.device_pixel_ratio,
+            None,
+            rendering_mode,
+            dwrote::DWRITE_MEASURING_MODE_NATURAL,
+            0.0,
+            0.0,
+        )
+        .map_err(Error::DirectWriteError)?;
+
+        let bounds = glyph_analysis
+            .get_alpha_texture_bounds(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1)
+            .map_err(Error::DirectWriteError)?;
+
+        let buf = glyph_analysis
+            .create_alpha_texture(dwrote::DWRITE_TEXTURE_CLEARTYPE_3x1, bounds)
+            .map_err(Error::DirectWriteError)?;
+
+        Ok(RasterizedGlyph {
+            c,
+            width: (bounds.right - bounds.left) as i32,
+            height: (bounds.bottom - bounds.top) as i32,
+            top: -bounds.top,
+            left: bounds.left,
+            buf: BitmapBuffer::RGB(buf),
+        })
+    }
+
+    fn get_loaded_font(&self, font_key: FontKey) -> Result<&Font, Error> {
+        self.fonts.get(&font_key).ok_or(Error::FontNotLoaded)
+    }
+
+    fn get_glyph_index(&self, face: &FontFace, c: char) -> Result<u16, Error> {
+        let idx = *face
+            .get_glyph_indices(&[c as u32])
+            .first()
+            // DirectWrite returns 0 if the glyph does not exist in the font
+            .filter(|glyph_index| **glyph_index != 0)
+            .ok_or_else(|| Error::MissingGlyph(c))?;
+
+        Ok(idx)
+    }
+
+    fn get_fallback_font(&self, loaded_font: &Font, c: char) -> Option<dwrote::Font> {
+        let fallback = self.fallback.as_ref()?;
+
+        let mut buf = [0u16; 2];
+        c.encode_utf16(&mut buf);
+
+        let length = c.len_utf16() as u32;
+        let utf16_codepoints = &buf[..length as usize];
+
+        let locale = get_current_locale();
+
+        let text_analysis_source_data = TextAnalysisSourceData { locale: &locale, length };
+        let text_analysis_source = TextAnalysisSource::from_text(
+            Box::new(text_analysis_source_data),
+            Cow::Borrowed(&utf16_codepoints),
```

Comment:
```suggestion
            Cow::Borrowed(utf16_codepoints),
```

I don't think this reference is necessary?